src/components/apps/launch/ParamGroups.js

```diff
@@ -0,0 +1,375 @@
+        default:
+            return baseParamId;
+    }
+};
+
+/**
+ * Form fields and info display for an app parameter group.
+ */
+const ParamGroupForm = withI18N((props) => {
+    const classes = useStyles();
+
+    const {
+        baseId,
+        fieldName,
+        group,
+        referenceGenomes,
+        referenceGenomesLoading,
+        startingPath,
+    } = props;
+
+    return (
+        <ExpansionPanel id={baseId} defaultExpanded>
+            <ExpansionPanelSummary
+                expandIcon={
+                    <ExpandMore id={buildDebugId(baseId, ids.BUTTONS.EXPAND)} />
+                }
+            >
+                <Typography variant="subtitle1">{group.label}</Typography>
+            </ExpansionPanelSummary>
+            <ExpansionPanelDetails className={classes.expansionPanelDetails}>
+                {group.parameters?.map((param, paramIndex) => {
+                    if (!param.isVisible) {
+                        return null;
+                    }
+
+                    const name = `${fieldName}.parameters.${paramIndex}.value`;
+                    const paramFormId = buildParamId(
+                        baseId,
+                        paramIndex,
+                        param.type
+                    );
+
+                    let fieldProps = {
+                        id: paramFormId,
+                        name,
+                        label: param.label,
+                        required: param.required,
+                    };
+
+                    switch (param.type) {
+                        case constants.PARAM_TYPE.INFO:
+                            fieldProps = {
+                                id: paramFormId,
+                                name,
+                                component: Typography,
+                                variant: "body1",
+                                gutterBottom: true,
+                                dangerouslySetInnerHTML: {
+                                    __html: sanitizeHtml(param.label),
+                                },
+                            };
+                            break;
+
+                        case constants.PARAM_TYPE.TEXT:
+                            fieldProps.component = FormTextField;
+
+                            if (param.validators?.length > 0) {
+                                const charLimitValidator = param.validators.find(
+                                    (validator) =>
+                                        validator.type ===
+                                        constants.VALIDATOR_TYPE.CHARACTER_LIMIT
+                                );
+                                if (charLimitValidator) {
+                                    fieldProps.inputProps = {
+                                        maxLength: charLimitValidator.params[0],
+                                    };
+                                }
+                            }
```

Comment:
This is the only spot where param validation is checked outside of the `validate` file, since `maxLength` can be set directly on a text field.